### PR TITLE
[Clang][OpenMP][Tile] Allow non-constant tile sizes.

### PR DESCRIPTION
**Pull request by Meinersbur**
Allow non-constants in the `sizes` clause such as
```
#pragma omp tile sizes(a)
for (int i = 0; i < n; ++i)
```
This is permitted since tile was introduced in [OpenMP 5.1](https://www.openmp.org/spec-html/5.1/openmpsu53.html#x78-860002.11.9).

It is possible to sneak-in negative numbers at runtime as in
```
int a = -1;
#pragma omp tile sizes(a)
```
Even though it is not well-formed, it should still result in every loop iteration to be executed exactly once, an invariant of the tile construct that we should ensure. `ParseOpenMPExprListClause` is extracted-out to be reused by the `permutation` clause if the `interchange` construct. Some care was put in to ensure correct behavior in template contexts.

This patch also adds end-to-end tests. This is to avoid errors like the off-by-one error I caused with the initial implementation of the unroll construct.

### CODE REVIEW

**Comment by Meinersbur** [comment]
Test failure is from unrelated `DataFlowSanitizer-x86_64 :: release_shadow_space.c`

File: clang/lib/Sema/SemaOpenMP.cpp [hunk 1]
```diff
@@ -15111,13 +15111,11 @@ StmtResult SemaOpenMP::ActOnOpenMPTileDirective(ArrayRef<OMPClause *> Clauses,
   ASTContext &Context = getASTContext();
   Scope *CurScope = SemaRef.getCurScope();
 
-  auto SizesClauses =
-      OMPExecutableDirective::getClausesOfKind<OMPSizesClause>(Clauses);
-  if (SizesClauses.empty()) {
-    // A missing 'sizes' clause is already reported by the parser.
+  const OMPSizesClause *SizesClause =
```
Comment (alexey-bataev) [inline clang/lib/Sema/SemaOpenMP.cpp:15114]:
```suggestion
  const auto *SizesClause =
```

File: clang/lib/Sema/SemaOpenMP.cpp [hunk 2]
```diff
@@ -15197,6 +15202,36 @@ StmtResult SemaOpenMP::ActOnOpenMPTileDirective(ArrayRef<OMPClause *> Clauses,
...
   Stmt *Inner = Body;
 
+  auto MakeDimTileSize = [&SemaRef = this->SemaRef, &CopyTransformer, &Context,
+                          SizesClause, CurScope](int I) -> Expr * {
+    Expr *DimTileSizeExpr = SizesClause->getSizesRefs()[I];
+    if (isa<ConstantExpr>(DimTileSizeExpr))
+      return AssertSuccess(CopyTransformer.TransformExpr(DimTileSizeExpr));
+
+    // When the tile size is not a constant but a variable, it is possible to
+    // pass non-positive numbers. To preserve the invariant that every loop
```
Comment (alexey-bataev) [inline clang/lib/Sema/SemaOpenMP.cpp:15212]:
```suggestion
    // pass positive numbers. To preserve the invariant that every loop
```

Comment (alexey-bataev) [reply clang/lib/Sema/SemaOpenMP.cpp:15212]: Or what do you mean here?

File: clang/lib/Sema/SemaOpenMP.cpp [hunk 3]
```diff
@@ -17432,16 +17457,54 @@ OMPClause *SemaOpenMP::ActOnOpenMPSizesClause(ArrayRef<Expr *> SizeExprs,
                                               SourceLocation StartLoc,
                                               SourceLocation LParenLoc,
                                               SourceLocation EndLoc) {
-  for (Expr *SizeExpr : SizeExprs) {
-    ExprResult NumForLoopsResult = VerifyPositiveIntegerConstantInClause(
-        SizeExpr, OMPC_sizes, /*StrictlyPositive=*/true);
-    if (!NumForLoopsResult.isUsable())
-      return nullptr;
+  SmallVector<Expr *> SanitizedSizeExprs;
+  llvm::append_range(SanitizedSizeExprs, SizeExprs);
```
Comment (alexey-bataev) [inline clang/lib/Sema/SemaOpenMP.cpp:17461]:
```suggestion
  SmallVector<Expr *> SanitizedSizeExprs(SizeExprs.begin(), SizeExprs.end());
```

File: clang/lib/Sema/SemaOpenMP.cpp [hunk 4]
```diff
@@ -17432,16 +17457,54 @@ OMPClause *SemaOpenMP::ActOnOpenMPSizesClause(ArrayRef<Expr *> SizeExprs,
...
+    //       Invalidate the expression with a clearly wrong value (nullptr) so
+    //       later in Sema we do not have to do the same validity analysis again
+    //       or crash from unexpected data. Error diagnostics have already been
+    //       emitted.
+    // 2b. The expression is invalid and we are in a template (including partial
+    //     instantiation):
+    //       Pass the invalid expression as-is, template instantiation may
+    //       replace unexpected types/values with valid ones. The directives
+    //       with this clause must not try to use these expressions in dependent
+    //       contexts.
```
Comment (alexey-bataev) [inline clang/lib/Sema/SemaOpenMP.cpp:17501]: This must be fixed, even if tricky

Comment (Meinersbur) [reply clang/lib/Sema/SemaOpenMP.cpp:17500]: Case 2b is already adhered to in `ActOnOpenMPTileDirective`:
```
  // Delay tiling to when template is completely instantiated.
  if (SemaRef.CurContext->isDependentContext())
    return OMPTileDirective::Create(Context, StartLoc, EndLoc, Clauses,
                                    NumLoops, AStmt, nullptr, nullptr);
```
Delaying further analysis seems generally to be what OpenMP does, e.g. 
```
isNonNegativeIntegerValue(...) {
  if (!ValExpr->isTypeDependent() && !ValExpr->isValueDependent() &&
      !ValExpr->isInstantiationDependent()) {
```

File: clang/lib/Sema/SemaOpenMP.cpp (hunk 2 above)
Comment (alexey-bataev) [reply clang/lib/Sema/SemaOpenMP.cpp:15212]: Sure. I'm just saying that it is better to clarify this comment.

File: clang/lib/Sema/SemaOpenMP.cpp (hunk 4 above)
Comment (alexey-bataev) [reply clang/lib/Sema/SemaOpenMP.cpp:17500]: Sure, delaying is a good approach. If the analysis is delayed till actual instantiation, it does not mean that the templates are not handled.

File: clang/lib/Sema/SemaOpenMP.cpp (hunk 2 above)
Comment (Meinersbur) [reply clang/lib/Sema/SemaOpenMP.cpp]: Added the example to the comment to clarify

File: clang/lib/Sema/SemaOpenMP.cpp (hunk 4 above)
Comment (Meinersbur) [reply clang/lib/Sema/SemaOpenMP.cpp]: Added clarification

**Review by alexey-bataev — APPROVED** [review]
LG

**Pull request merged**